src/ising.2.cpp

```diff
@@ -0,0 +1,245 @@
+    
+    double beta = 1.0/T;
+    int * point = new int [3];
+    int i = (int) rnd()%N; int j = (int)rnd()%N; int k = (int) rnd()%N;
+    
+    point[0] = i; point[1] = j; point[2]=k;
+    
+    
+    double dE = deltaE(N, si, h, J, point);
+    double dm=0.0,dh=0.0;
+    double p = T*log(rnd());
+    if (dE > p)
+    {
+        
+        si[i][j][k] *= -1;
+        dm=2.0*si[i][j][k];
+        dh=dE;
+    }
+    
+    M = M+dm;
+    H = H+dh;
+    
+    free (point);
+    return;
+}
+
+void collect_data(double * data, double M, double H){
+    data[0] = data[0]+M;
+    data[1] = data[1]+H;
+    data[2] = data[2]+M*M;
+    data[3] = data[3]+H*H;
+    data[4] = data[4]+fabs(M);
+}
+
+void configuration(int N, int *** si, string filename){
+    
+    ofstream cfile;
+    cfile.open(filename.c_str());
+    
+    for (int i = 0; i < N; i++)
+        for (int j = 0; j < N; j++)
+            for (int k = 0; k < N; k++)
+            {
+                cfile<<i <<"\t" << j << "\t" << k << "\t" << si[i][j][k] << "\n";
+            }
+            
+            cfile.close();
+        return;
+    
+}
+
+int main(){
+   
+    
+  int N = 10;
+  int MCSteps = 300*N*N*N;
+  double h = 1.0;
+  double J = 2.0;
+  
+  int *** si = new int ** [N];
+  initialize(N, si);
+  
+  double T;
+  double M = Magnetization(N, si);
+  double H = Energy(N, si, h, J);
+  
+  // do{
+        cout << "Running h = " << h << "\n";
+        ofstream tfile;
+        std::ostringstream tf;
+        tf << h <<"_mg.txt" ;
+        std::string tmf = tf.str();
+        tfile.open(tmf.c_str());
+        
+        
+        for (T=1.0; T <= 10.; T+=1.){
+            
+            
+            double * data;
+            data = new double [5];
+            for (int i=0; i < 5; i++) data[i] = 0.0;
+            
+            
+            //Equilibration Steps
+            for (int step = 0; step < MCSteps; step++)
+                MC_move(N, si, h, J, M, H, T);
+            
+            //Production
+            for (int step = 0; step < 0.5*MCSteps; step++){
+                
+                MC_move(N, si, h, J, M, H, T);
+                
+                //recording and updating the M and H values 
+                //Frequency can be varied depending on length/size of simulation
+                collect_data(data, M, H);
+            }
+            
+            //Normalization
+            double norm   = 1.0/(MCSteps*N*N*N);
+            double M_avg  = data[0]*norm;
+            double H_avg  = data[1]*norm;
+            double M_2avg = data[2]*norm;
+            double H_2avg = data[3]*norm;
+            double M_abs  = data[4]*norm;
+            
+            
+            //Quantities
+            
+            double beta = 1.0/T;
+            double beta2 = beta*beta;
+            
+            //Susceptibility
+            double X = beta*( M_2avg - (M_avg*M_avg));
+            //Specific Heat
+            double Cv = beta2*(H_2avg - (H_avg*H_avg));
```

Comment:
Specific heat should be
```cpp
double Cv = (beta2 / (N * N * N)) * (H_2avg - (H_avg * H_avg));
```
provided the `norm` variable is fixed, see above comment.